src/response.rs

```diff
@@ -180,9 +181,17 @@ impl Response {
         let len = self.headers().get::<::header::ContentLength>()
             .map(|ct_len| **ct_len)
             .unwrap_or(0);
-        let mut content = String::with_capacity(len as usize);
-        self.read_to_string(&mut content).map_err(::error::from)?;
-        Ok(content)
+        let mut content = Vec::with_capacity(len as usize);
+        self.read_to_end(&mut content).map_err(::error::from)?;
+        let encoding_name = self.headers().get::<::header::ContentType>()
+            .and_then(|content_type| {
+                content_type.get_param("charset")
+                    .map(|charset| charset.as_str().to_string())
+            })
+            .unwrap_or_else(|| "utf-8".to_string());
+        let encoding = Encoding::for_label(encoding_name.as_bytes()).unwrap_or(UTF_8);
+        let (text, _, _) = encoding.decode(&content);
```

Comment:
It looks like [`decode`](https://docs.rs/encoding_rs/0.7.*/encoding_rs/struct.Encoding.html#method.decode) returns a `Cow<str>`, since it may have detected that the bytes were valid UTF-8 and didn't need to do any copying. So, we can handle the `Cow` if it is `Cow::Borrowed`, that means we don't need to make a new copy, since the bytes in `content` were valid! Eliminating this copy is a bigger deal depending on how big the body was.

So, seems this could be handled like so:

```rust
// a block because of borrow checker
{
    let (text, _, _) = encoding.decode(&content);
    match text {
        Cow::Owned(s) => return Ok(s),
        _ => (),
    }
}
unsafe {
    // decoding returned Cow::Borrowed, meaning these bytes
    // are already valid utf8
    Ok(String::from_utf8_unchecked(content))
}
```